storage/remote/stream.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package remote
+
+import (
+	"bufio"
+	"encoding/binary"
+	"github.com/gogo/protobuf/proto"
+	"io"
+)
+
+// StreamWriter is an io.Writer wrapper that allows streaming by adding uvarint delimiter before each write in a form
+// of length of the corresponded byte array.
+type StreamWriter struct {
+	wrapped io.Writer
+}
+
+// NewStreamWriter constructs a StreamWriter.
+func NewStreamWriter(w io.Writer) *StreamWriter {
+	return &StreamWriter{wrapped: w}
+}
+
+// Write writes given bytes to the stream. It adds uvarint delimiter before each message.
+// Returned bytes number represents sent bytes for a given buffer. The number does not include delimiter bytes.
+func (w *StreamWriter) Write(b []byte) (int, error) {
+	if len(b) == 0 {
+		return 0, nil
+	}
+
+	var buf [binary.MaxVarintLen64]byte
+	v := binary.PutUvarint(buf[:], uint64(len(b)))
+
+	if _, err := w.wrapped.Write(buf[:v]); err != nil {
+		return 0, err
+	}
+
+	n, err := w.wrapped.Write(b)
+	if err != nil {
+		return n, err
+	}
+
+	return n, nil
```

Comment:
Its hard to know without seeing the entire implementation, but I think you'll need a flush to ensure the data isn't buffered in the server.

Reply:
Yes. I planned to do it on caller side, but happy to ensure this here. Let's see what is more clear.